src/compiler/checker.ts

```diff
@@ -37014,7 +37014,10 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             if (!witnesses) {
                 return false;
             }
-            const operandConstraint = getBaseConstraintOrType(checkExpressionCached((node.expression as TypeOfExpression).expression));
+            const type = checkExpressionCached((node.expression as TypeOfExpression).expression);
+            const operandConstraint = type.flags & (TypeFlags.Intersection | TypeFlags.Instantiable) ?

```

Comment:
Why do we want to do this when the type is intersection or instantiable?

Reply:
Doing this for instantiables shouldn't fall into the trap that you have mentioned (with primitive types). I'm not sure about intersections though, I just followed the same logic that I found at the top of `getTypeFactsWorker`.

Reply:
I see.
I was going to say I don't know exactly if this is correct, but I realized there might be a simpler way to get what we want:
we can always say that the `typeof` switch is exhaustive if it has all possible `typeof` strings, regardless of the operand constraint.
So I think we can leave this all unchanged and just check if `(TypeFacts.AllTypeofNE & notEqualFacts) === TypeFacts.AllTypeofNE` is true, and if it is, we can return true (and otherwise we can do what we were already doing).